File: benchmarks/java/set/Java_Persistent/resources/BRelation.java
package de.hhu.stups.btypes;

import clojure.lang.PersistentHashSet;

import java.util.List;
import java.util.stream.Collectors;

/**
 * Created by fabian on 15.01.19.
 */
public class BRelation<S,T> extends BSet<BCouple<S,T>> {

    @SafeVarargs
    public BRelation(BCouple<S,T>... elements) {
        super(elements);
    }

    public BSet<T> relationImage(BSet<S> domain) {
        return new BSet(HashTreePSet.from(set.stream()
                .filter(object -> domain.contains(((BCouple) object).getFirst()))
                .map(object -> ((BCouple) object).getSecond())
                .collect(Collectors.toSet())));
    }


    public T functionCall(BObject arg) {
        List<T> matchedCouples = set.stream()
                .map(object -> (BCouple) object)
                .filter(couple -> couple.getFirst().equals(arg))
                .collect(Collectors.toList());
        if(matchedCouples.size() > 0) {
            return matchedCouples.get(0).getSecond();
        }
        throw new RuntimeException("Argument is not in the key set of this map");
    }

}

File: src/test/resources/de/hhu/stups/codegenerator/recursion/BRelation.cpp
#include <iostream>
#include <string>
#include <cstdarg>
#include "BSet.cpp"
#include "BCouple.cpp"

#ifndef BRELATION_H
#define BRELATION_H

using namespace std;

template<typename S, typename T>
class BRelation<S,T> : public BSet<BCouple<S,T>> {

    BSet<T> relationImage(const BSet<S>& domain) {
        immer::set<T,Hash, HashEqual> result;
        for(typename immer::set<BCouple<S,T>,Hash, HashEqual>::const_iterator it = this->set.begin(); it != this->set.end(); ++it) {
            BCouple<S,T> couple = *it;
            if(domain.set.count(couple.getFirst()) == 0) {
                result = result.insert(couple.getSecond());
            }
        }
        return BSet(result);
    }


    T functionCall(const S& arg) {
        for(typename immer::set<<BCouple<S,T>>,Hash, HashEqual>::const_iterator it = this->set.begin(); it != this->set.end(); ++it) {
            BCouple<S,T> couple = *it;
            if(couple.getFirst() == arg) {
                return couple.getSecond();
            }
        }
        throw runtime_error("Argument is not in the key set of this map");
    }

}